bitcoin/src/witness_program.rs

```diff
@@ -55,64 +80,285 @@ impl WitnessProgram {
-            return Err("Invalid program length");
+            return Err(WitnessProgramError::InvalidProgramLength(self.program.len()));
         }
 
         if self.version > 16 {
-            return Err("Invalid version");
+            return Err(WitnessProgramError::InvalidVersion(self.version));
         }
 
         // P2SH_P2WPKH start with 0x0014
         // P2SH_P2WSH starts with 0x0020
         // https://bitcoincore.org/en/segwit_wallet_dev/#creation-of-p2sh-p2wpkh-address
         if self.version == 0 && !(self.program.len() == 20 || self.program.len() == 32) {
-            return Err("Invalid program length for witness version 0");
+            return Err(WitnessProgramError::InvalidProgramLengthForVersion(self.program.len(), self.version));
         }
 
         Ok(())
     }
 
-    /// Returns the witness program as a byte vector.
-    pub fn to_vec(&self) -> Vec<u8> {
-        let mut output = vec![WitnessProgram::convert_version(self.version)];
+    /// Returns the witness program's scriptpubkey as a byte vector.
+    pub fn to_scriptpubkey(&self) -> Vec<u8> {
+        let mut output = Vec::with_capacity(self.program.len() + 2);
+        let encoded_version = if self.version > 0 {
+            self.version + 0x50
+        } else {
+            self.version
+        };
+        output.push(encoded_version);
         output.push(self.program.len() as u8);
-        output.copy_from_slice(&self.program);
+        output.extend_from_slice(&self.program);
         output
     }
-
-    /// A BIP173 version conversion utility function.
-    /// Convert a given version to a value between 0 and 16 or OP_1 through OP_16.
-    pub fn convert_version(version: u8) -> u8 {
-        if version > 0x00 && version <= 0x10 {
-            // encode OP_1 through OP_16
-            version + 0x50
-        } else if version > 0x50 {
-            // decode OP_1 through OP_16
-            version - 0x50
-        } else {
-            // OP_0
-            version
-        }
-    }
 }
 
 impl FromStr for WitnessProgram {
-    type Err = &'static str;
+    type Err = WitnessProgramError;
 
     /// Returns a witness program given its hex representation.
     fn from_str(s: &str) -> Result<Self, Self::Err> {
         WitnessProgram::new(&match hex::decode(s) {
             Ok(bytes) => bytes,
-            Err(_) => return Err("Error decoding hex string"),
+            Err(_) => return Err(WitnessProgramError::ProgramDecodingError),
         })
     }
 }
 
-impl fmt::Display for WitnessProgram {
-    /// Prints a witness program in hex representation.
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", hex::encode(self.to_vec()))
+#[cfg(test)]
+mod tests {
+    mod test_from_str {
+        use std::str::FromStr;
+
+        use crate::witness_program::{WitnessProgram, WitnessProgramError};
+
+        const VALID_P2SH_P2WPKH_PROGRAM: &str = "0014751e76e8199196d454941c45d1b3a323f1433bd6";
```

Comment:
Adhere to the test suite convention of providing a slice containing all programs at the top of the inner mod.

See https://github.com/ArgusHQ/wagyu/blob/v0.6.0/bitcoin/src/extended_private_key.rs#L682